### ISSUE TITLE
Build failure with GHC 9.2.1

### ISSUE THREAD

**Issue by sjakobi** [Lysxia/generic-random#32]
```
src/Generic/Random/Internal/BaseCase.hs:191:3: error:
    • Could not deduce (Alternative (IfM y Weighted Proxy))
        arising from the superclasses of an instance declaration
      from the context: (GBCSSum f g z e yf yg, GBCS f z yf e,
                         GBCS g z yg e, y ~ (yf ||? yg))
        bound by the instance declaration
        at src/Generic/Random/Internal/BaseCase.hs:(191,3)-(195,27)
    • In the instance declaration for ‘GBCS (f :+: g) z y e’
    |
191 |   ( GBCSSum f g z e yf yg
    |   ^^^^^^^^^^^^^^^^^^^^^^^...

src/Generic/Random/Internal/BaseCase.hs:239:3: error:
    • Could not deduce (Alternative (IfM y Weighted Proxy))
        arising from the superclasses of an instance declaration
      from the context: (GBCSProduct f g z e yf yg, GBCS f z yf e,
                         GBCS g z yg e, y ~ (yf &&? yg))
        bound by the instance declaration
        at src/Generic/Random/Internal/BaseCase.hs:(239,3)-(243,27)
    • In the instance declaration for ‘GBCS (f :*: g) z y e’
    |
239 |   ( GBCSProduct f g z e yf yg
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^...
cabal: Failed to build generic-random-1.5.0.0
```

**Comment by sjakobi** [comment]
I'm getting the same error with all the versions from v1.2.0.0 and up.

I'm wondering whether I should add Hackage revisions to add bounds `base < 4.16` for these versions, but I'm actually not sure whether the problem is in this library or in GHC.

**Comment by Lysxia** [comment]
Huh, that seems to be a regression coming from GHC. In any case it wouldn't hurt to add the bounds to the released versions that broke, would it?

**Comment by Lysxia** [comment]
I opened a GHC ticket with a minimized example https://gitlab.haskell.org/ghc/ghc/-/issues/20666

**Comment by sjakobi** [comment]
Thanks!

I have made the Hackage revisions, e.g. https://hackage.haskell.org/package/generic-random-1.5.0.0/revisions/.

Once there's fixed GHC 9.2.x, we can probably relax these again.

**Comment by goldfirere** [comment]
#34 contains the fix.

GHC uses a sizing scheme to ensure that it does not create bottom dictionaries. When trying to verify that an instance's superclass constraints are satisfied, it will use superclass constraints extracted from *smaller* instance context constraints. In the two instance declarations patched in #34, both the relevant instance constraint and the instance head (the part to the right of the `=>`) have size 6, so GHC won't use the superclass of the instance constraint. Maybe it's possible to do some syntactic manipulation to get the sizes to be different, but there seems little harm in just adding the redundant constraint (which is, happily, backward compatible).

The fact that older GHCs accepted the instances was a bug. I'm fairly sure that it would have been possible to create bottom dictionaries.

**Comment by Lysxia** [comment]
Thanks for the detailed explanation Richard!

@sjakobi I released the fix in 1.5.0.1

**Comment by sjakobi** [comment]
Thank you, @Lysxia and @goldfirere! :)
